source/bigint.c

```diff
@@ -0,0 +1,353 @@
+    if (hex_cursor->len >= 2) {
+        const char *raw_ptr = (char *)hex_cursor->ptr;
+        if (raw_ptr[0] == '0' && (raw_ptr[1] == 'x' || raw_ptr[1] == 'X')) {
+            aws_byte_cursor_advance(hex_cursor, 2);
+        }
+
+        while (hex_cursor->len > 0 && *hex_cursor->ptr == '0') {
+            aws_byte_cursor_advance(hex_cursor, 1);
+        }
+    } else if (hex_cursor->len == 1) {
+        if (*hex_cursor->ptr == '0') {
+            aws_byte_cursor_advance(hex_cursor, 1);
+        }
+    }
+}
+
+static int s_uint32_from_hex(struct aws_byte_cursor digit_cursor, uint32_t *output_value) {
+    AWS_FATAL_ASSERT(digit_cursor.len <= NIBBLES_PER_DIGIT);
+
+    *output_value = 0;
+
+    while (digit_cursor.len > 0) {
+        char hex_digit = *digit_cursor.ptr;
+        uint32_t hex_value = 0;
+        if (hex_digit <= '9' && hex_digit >= '0') {
+            hex_value = hex_digit - '0';
+        } else if (hex_digit <= 'f' && hex_digit >= 'a') {
+            hex_value = hex_digit - 'a' + 10;
+        } else if (hex_digit <= 'F' && hex_digit >= 'A') {
+            hex_value = hex_digit - 'A' + 10;
+        } else {
+            return AWS_OP_ERR;
+        }
+
+        *output_value <<= 4;
+        *output_value += hex_value;
+
+        aws_byte_cursor_advance(&digit_cursor, 1);
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+int aws_bigint_init_from_hex(
+    struct aws_bigint *bigint,
+    struct aws_allocator *allocator,
+    struct aws_byte_cursor hex_digits) {
+    s_advance_cursor_to_hex_start(&hex_digits);
+
+    if (hex_digits.len == 0) {
+        return aws_bigint_init(bigint, allocator);
+    }
+
+    uint64_t digit_count = hex_digits.len / BASE_BITS + 1;
+    if (aws_array_list_init_dynamic(&bigint->digits, allocator, digit_count, sizeof(uint32_t))) {
+        return AWS_OP_ERR;
+    }
+
+    while (hex_digits.len > 0) {
+        struct aws_byte_cursor digit_cursor = hex_digits;
+        if (digit_cursor.len > NIBBLES_PER_DIGIT) {
+            digit_cursor.ptr += (digit_cursor.len - NIBBLES_PER_DIGIT);
+            digit_cursor.len = NIBBLES_PER_DIGIT;
+        }
+
+        uint32_t digit_value = 0;
+        if (s_uint32_from_hex(digit_cursor, &digit_value)) {
+            goto on_error;
+        }
+
+        aws_array_list_push_back(&bigint->digits, &digit_value);
+
+        hex_digits.len -= digit_cursor.len;
+    }
+
+    bigint->sign = 1;
+
+    return AWS_OP_SUCCESS;
+
+on_error:
+
+    aws_bigint_clean_up(bigint);
+
+    return AWS_OP_ERR;
+}
+
+int aws_bigint_init_from_int64(struct aws_bigint *bigint, struct aws_allocator *allocator, int64_t value) {
+    if (value >= 0) {
+        return aws_bigint_init_from_uint64(bigint, allocator, (uint64_t)value);
+    }
+
+    if (value == INT64_MIN) {
+        if (aws_bigint_init_from_uint64(bigint, allocator, INT64_MIN_AS_HEX)) {
+            return AWS_OP_ERR;
+        }
+    } else {
+        if (aws_bigint_init_from_uint64(bigint, allocator, (uint64_t)(-value))) {
```

Comment:
I don't know that I understand why the negation is here, comment maybe?

Reply:
Sure, just ensuring that negating the negative integer would be representable as a positive integer (pre uint64 cast).  Maybe that kind of care is not necessary if you understand the C standard and defined/undefined behavior better than I.

Will add comments.